libvmaf/test/test_feature.c

```diff
@@ -49,8 +50,137 @@ static char *test_feature_name()
+            .default_val.d = opt_double_default,
+        },
+        {
+            .name = "opt_int",
+            .alias = "opt_int_alias",
+            .offset = offsetof(TestState, opt_int),
+            .type = VMAF_OPT_TYPE_INT,
+            .default_val.i = opt_int_default,
+        },
+        { 0 }
+    };
+
+    TestState s1 = {
+        .opt_bool = opt_bool_default,
+        .opt_double = opt_double_default,
+        .opt_int = opt_int_default,
+    };
+
+    char *feature_name1 =
+        vmaf_feature_name_from_options("feature_name", &options, &s1, 3,
+                                       &s1.opt_bool, &s1.opt_double, &s1.opt_int);
+
+    mu_assert("when all options are default, feature_name should not change",
+              !strcmp(feature_name1, "feature_name"));
+
+    free(feature_name1);
+
+    TestState s2 = {
+        .opt_bool = !opt_bool_default,
+        .opt_double = opt_double_default,
+        .opt_int = opt_int_default,
+    };
+
+    char *feature_name2 =
+        vmaf_feature_name_from_options("feature_name", &options, &s2, 3,
+                                       &s2.opt_bool, &s2.opt_double, &s2.opt_int);
+
+    mu_assert("when opt_bool has a non-default value, "
+              "feature_name should have a non-aliased opt_bool suffix",
+              !strcmp(feature_name2, "feature_name_opt_bool_1"));
+
+    free(feature_name2);
+
+    TestState s3 = {
+        .opt_bool = opt_bool_default,
+        .opt_double = opt_double_default + 1,
+        .opt_int = opt_int_default,
+    };
+
+    char *feature_name3 =
+        vmaf_feature_name_from_options("feature_name", &options, &s3, 3,
+                                       &s3.opt_bool, &s3.opt_double, &s3.opt_int);
+
+    mu_assert("when opt_double has a non-default value, "
+              "feature_name should have a aliased opt_double_alias suffix",
+              !strcmp(feature_name3, "feature_name_opt_double_alias_4.14"));
+
+    free(feature_name3);
+
+    TestState s4 = {
+        .opt_bool = !opt_bool_default,
+        .opt_double = opt_double_default + 1,
+        .opt_int = opt_int_default + 1,
+    };
+
+    char *feature_name4 =
+        vmaf_feature_name_from_options("feature_name", &options, &s4, 3,
+                                       &s4.opt_bool, &s4.opt_double, &s4.opt_int);
+
+
+    mu_assert("when all opts have a non-default value, "
+              "feature_name should have a suffix with aliases and values",
+              !strcmp(feature_name4, "feature_name_opt_bool_1_opt_double_alias_4.14_opt_int_alias_201"));
+
+    free(feature_name4);
+
+    TestState s5 = s4;
+
+    char *feature_name5 =
+        vmaf_feature_name_from_options("feature_name", &options, &s5, 3,
+                                       &s5.opt_double, &s5.opt_bool, &s5.opt_int);
+
+    mu_assert("feature_name should have a suffix with aliases and values, "
+              "ordering should follow the ordering of variadac params",
```

Comment:
"ordering should follow the ordering of variadac params" ==> this could be a source of bug. We would want to use the feature_name as a unique identifier, but in case a user calls vmaf_feature_name_from_options twice with different options ordering, it will result in two feature names, although the feature numerical results are the same. Two possible mitigations: either 1) check to see if the input options are sorted alphabetically; if not, return an error code; 2) sort the input options before constructing the feature name string.